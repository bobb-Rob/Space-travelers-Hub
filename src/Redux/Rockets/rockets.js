import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const getSelectField = (rockets) => {
  let newRockets = [];
  rockets.forEach((rocket) => {
    const selectedField = {};
    selectedField.id = rocket.id;
    selectedField.name = rocket.rocket_name;
    selectedField.type = rocket.rocket_type;
    const [imageOne] = rocket.flickr_images;
    selectedField.flickr_images = imageOne;
    newRockets = [...newRockets, selectedField];
  });
  return newRockets;
};

export const getRockets = createAsyncThunk(
  'rockets/getRockets',
  async () => {
    const response = await fetch('https://api.spacexdata.com/v3/rockets');
    const rockets = await response.json();
    return getSelectField(rockets);
  },
);

const initialState = [];

const rocketSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {},
  extraReducers: {
    [getRockets.fulfilled]: (state, action) => [...state, ...action.payload],
  },
});

export default rocketSlice.reducer;
