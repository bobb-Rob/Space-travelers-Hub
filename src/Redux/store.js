import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import missionReducer from './mission';

const store = configureStore({
  reducer: {
    missionReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
});

export default store;
