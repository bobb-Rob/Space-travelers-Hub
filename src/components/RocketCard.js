import PropTypes from 'prop-types';

const RocketCard = ({ description, name, image }) => (
  <div className="rocket-card">
    <div>
      <img src={image} alt={name} />
    </div>
    <div>
      <h3>{name}</h3>
      <p>{description}</p>
      <button type="button">Reserve Rocket</button>
    </div>
  </div>
);

export default RocketCard;

RocketCard.propTypes = {
  description: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
};
