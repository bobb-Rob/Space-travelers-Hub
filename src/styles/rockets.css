.rockets-list-container {
  font-family: Montserrat, sans-serif;
  font-weight: 300;
  font-stretch: normal;
  font-style: normal;
  line-height: normal;
  letter-spacing: normal;
}

.rocket-card {
  display: grid;
  grid-template-columns: 1fr 3fr;
  gap: 1.5rem;
  margin-bottom: 1.5rem;
}

.rocket-card div:first-child > img {
  width: 100%;
}

.card-details > h3,
.card-details > p {
  margin-bottom: 1rem;
}

.card-details > button {
  min-width: 150px;
  height: 2rem;
  border-radius: 4px;
  color: var(--white);
  background-color: var(--blue);
  border: none;
}
