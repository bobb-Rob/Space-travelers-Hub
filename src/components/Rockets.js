import '../styles/rockets.css';

const Rockets = () => (
  <div className="rockets-list-container">Rockets</div>
);

export default Rockets;
