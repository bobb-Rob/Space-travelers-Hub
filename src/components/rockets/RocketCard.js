import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserved } from '../../Redux/Rockets/rockets';

const RocketCard = ({
  description, name, image, id, status,
}) => {
  const dispatch = useDispatch();

  return (
    <div className="rocket-card">
      <div>
        <img src={image} alt={name} />
      </div>
      <div className="card-details">
        <h3>{name}</h3>
        <p>
          {status
            ? <span className="reserveTag">Reserved</span>
            : null}
          {' '}
          {description}
        </p>
        <button
          className={status ? 'reserved' : 'noreserved'}
          type="button"
          onClick={() => dispatch(reserved(id))}
        >
          {status ? 'Cancel Reservation' : 'Reserve Rocket'}
        </button>
      </div>
    </div>
  );
};

export default RocketCard;

RocketCard.propTypes = {
  description: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  status: PropTypes.bool.isRequired,
};
