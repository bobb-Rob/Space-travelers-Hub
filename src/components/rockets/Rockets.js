import { useSelector } from 'react-redux';
import RocketCard from './RocketCard';
import '../../styles/rockets.css';

const Rockets = () => {
  const state = useSelector((state) => state.rocketReducer);

  return (
    <div className="rockets-list-container">
      {state.map((rocket) => {
        const {
          description, rocketName, flickImage, id, status,
        } = rocket;

        return (
          <RocketCard
            key={id}
            description={description}
            image={flickImage}
            name={rocketName}
            id={id}
            status={status}
          />
        );
      })}
    </div>
  );
};
export default Rockets;
