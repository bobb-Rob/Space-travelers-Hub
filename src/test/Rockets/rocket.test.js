import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../Redux/store';
import { reserved } from '../../Redux/Rockets/rockets';
import Rockets from '../../components/rockets/Rockets';
import RocketCard from '../../components/rockets/RocketCard';

describe('should render with given state from Redux store', () => {
  test('Rockets.js renders correctly', () => {
    let tree = renderer
    .create(
    <Provider store={store}>
      <Rockets>
        <RocketCard />
      </Rockets>
    </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
})


describe('My connected react-redux commponents', ( )=> {  
  
  store.dispatch = jest.fn();

  test('It should dispatch an action', () => {
    render(
      <Provider store={store}>       
          <RocketCard />        
      </Provider>
    );  

    let element = screen.getByRole('button');     
    fireEvent.click(element);

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(reserved());
  })
});