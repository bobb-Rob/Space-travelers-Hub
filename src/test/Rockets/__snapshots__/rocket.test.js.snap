// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with given state from Redux store Rockets.js renders correctly 1`] = `
<div
  className="rockets-list-container"
/>
`;
